modin/core/dataframe/pandas/dataframe/dataframe.py

```diff
@@ -363,7 +416,10 @@ def _get_index(self):
         pandas.Index
             An index object containing the row labels.
         """
-        if self._index_cache is None:
+        if callable(self._index_cache):
+            self._index_cache, self._row_lengths_cache = self._index_cache()
```

Comment:
Should we add `if self._row_lengths_cache is None:` before setting the row lengths as in the branch below?